src/terms/operators.jl

```diff
@@ -96,12 +96,12 @@ Nonlocal operator in Fourier space in Kleinman-Bylander format,
 defined by its projectors P matrix and coupling terms D:
 Hψ = PDP' ψ
 """
-struct NonlocalOperator <: RealFourierOperator
-    basis
-    kpoint
+struct NonlocalOperator{T <: Real, PT, DT} <: RealFourierOperator
+    basis::PlaneWaveBasis{T}
+    kpoint::Kpoint{T}
     # not typed, can be anything that supports PDP'ψ
-    P
-    D
+    P::PT
+    D::DT
```

Comment:
On this and on others: why type the fields? This is not performance critical at all, and doesn't really help with readability (at least not these two)

Reply:
It does not really make it a lot worse in terms of readability in my opinion and at the same time downstream types (e.g. when computing P * D * P' * \psi can be fully inferred.

Reply:
Inference does not matter at all here. My experience is that you can get into big messes because of eager typing of everything. Not particularly critical here, but not a good practice in my opinion

Reply:
In this case I don't see the mess. It is a small struct with two fields to type-stably store some intermediate data. It's use is very local, the overhead in constructing and using it is zero due to this change. Has basically no downsides and only benefits (which might be small, but ok).

Reply:
Oh it's just my general heuristics of more characters = bad :-p Just makes it a bit harder to parse, and generic structs are slightly nastier (eg to print)

Reply:
Hmm. I have to say in the cases we have here I still lean to have the type annotations. It's not really make the code a lot longer or more complex, but it does give a bit of speed (ok arguably not in a part of the code it really matters, but energy_hamiltonian becomes notably faster).

Reply:
My general rule is that type annotations (when done for performance and not for clarity or safety) are an optimization, and *any* optimization has a (sometimes hidden) cost (more things to parse when reading code, more potential bugs, less flexibility if we want to change something, more inference time, etc), so optimizations are only to be done if profiling demonstrates they are relevant. We can keep it if you really want but I just want to avoid it being a pattern

Reply:
Ok, that's fair.